expat/tests/runtests.c

```diff
@@ -11975,6 +12022,209 @@ START_TEST(test_nsalloc_prefixed_element)
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "'>\n"
+        "]>\n"
+        "<he>&large;</he>\n";
+    expect_failure(text, XML_ERROR_ENTITY_VIOLATION_SIZE,
+                   "XML_ERROR_ENTITY_VIOLATION_SIZE not raised");
+}
+END_TEST
+
+START_TEST(test_huge_xml_entity_expansion_limit)
+{
+    const char *text =
+        "<!DOCTYPE he ["
+        "<!ELEMENT he (#PCDATA)*>"
+        "<!ENTITY e '"
+        "ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP"
+        "'>]><he>"
+        /* 50 + 64 + 8 + (36 * 3) == 230
+         * 36 expansions of 64 bytes == 2304 */
+        "&e;&e;&e;&e;&e;&e;&e;&e;&e;&e;"
+        "&e;&e;&e;&e;&e;&e;&e;&e;&e;&e;"
+        "&e;&e;&e;&e;&e;&e;&e;&e;&e;&e;"
+        "&e;&e;&e;&e;&e;&e;"
+        "</he>\n";
+    expect_failure(text, XML_ERROR_ENTITY_VIOLATION_RATIO,
+                   "XML_ERROR_ENTITY_VIOLATION_RATIO not raised");
+}
+END_TEST
+
+START_TEST(test_huge_xml_hash_table_limit)
+{
+    XML_Size tableEntries = 4; /* smaller than 4 entities + 1 element */
+    const char *text =
+        "<!DOCTYPE he [\n"
+        "  <!ELEMENT he (#PCDATA)*>\n"
+        "  <!ENTITY e1 '&e2;'>\n"
+        "  <!ENTITY e2 '&e3;'>\n"
+        "  <!ENTITY e3 '&e4;'>\n"
+        "  <!ENTITY e4 'entity'>\n"
+        "]>\n"
+        "<he>&e1;</he>\n";
+
+    if (XML_SetOption(parser, XML_OPTION_MAX_HASH_TABLE_ENTRIES, &tableEntries) != XML_STATUS_OK)
+        fail("XML_SetOptions(XML_OPTION_MAX_HASH_TABLE_ENTRIES) failed");
+
+    expect_failure(text, XML_ERROR_HASH_TABLE_VIOLATION,
+                   "XML_ERROR_HASH_TABLE_VIOLATION not raised");
+}
+END_TEST
+
+START_TEST(test_huge_xml_enabled)
+{
+    XML_Size smallSize = 1;
+    int smallInt = 1;
+    XML_Bool hugeXML = XML_TRUE;
+    const char *text =
+        "<!DOCTYPE he [\n"
+        "  <!ELEMENT he (#PCDATA)*>\n"
+        "  <!ENTITY e1 '&e2;'>\n"
+        "  <!ENTITY e2 '&e3;'>\n"
+        "  <!ENTITY e3 '&e4;'>\n"
+        "  <!ENTITY e4 'entity'>\n"
+        "]>\n"
+        "<he>&e1;</he>\n";
+
+    if (XML_SetOption(parser, XML_OPTION_MAX_HASH_TABLE_ENTRIES, &smallSize) != XML_STATUS_OK)
+        fail("XML_SetOptions(XML_OPTION_MAX_HASH_TABLE_ENTRIES) failed");
+    if (XML_SetOption(parser, XML_OPTION_NESTING_LIMIT, &smallInt) != XML_STATUS_OK)
+        fail("XML_SetOptions(XML_OPTION_NESTING_LIMIT) failed");
```

Comment:
To be complete, it would be nice to also cover `XML_OPTION_EXPANSION_RATIO` and `XML_OPTION_MAX_EXPANSION_SIZE` here.